### PR TITLE
Feat decorators

### PR DESCRIPTION
**Pull request by pepoviola**
Add _decorators_ logic to allow chain customization by parachains. Each parachain can add his customization by overriding one of the `fns` that customize the chain-spec

```
    addAuraAuthority,
    addAuthority,
    changeGenesisConfig,
    clearAuthorities,
    readAndParseChainSpec,
    specHaveSessionsKeys,
    getNodeKey,
    addParaCustom,
    addCollatorSelection,
    writeChainSpec,
 ```

by adding the parachain in the `enum` (in `para-decorators/index.ts`)
 
```
enum PARA {
    Statemint = "statemint",
    Moonbeam = "moonbeam",
    Generic = "generic"
   }
```

Then import the file with the overrides _fns_
```
// imports
import moonbeam from "./moonbeam";
import statemint from "./statemint";
```

And  finally add to the registry

```
const statemintDecorators: ParaDecorator = Object.keys(statemint).reduce((memo, fn) => {
    memo[fn] = (statemint as ParaDecorator)[fn];
    return memo;
},Object.create({}));

const decorators : {[para in PARA]: { [fn: string]: Function}} = {
    moonbeam : moonbeamDecorators,
    statemint: statemintDecorators,
    generic: {}
};
```

cc @crystalin, I moved the moonbeam specific code and added a smoke test to ensure block production.

### CODE REVIEW

File: src/chain-spec.ts [hunk 1]
```diff
@@ -114,86 +100,109 @@ export async function addBalances(specPath: string, nodes: Node[]) {
...
+      babe: sr_account.address,
+      im_online: sr_account.address,
+      parachain_validator: sr_account.address,
+      authority_discovery: sr_account.address,
+      para_validator: sr_account.address,
+      para_assignment: sr_account.address,
+      beefy: encodeAddress(ec_account.publicKey),
+      aura: sr_account.address,
+    },
+  ];
```
Comment (wirednkod) [inline src/chain-spec.ts:124]:
```suggestion
const address =  useStash ? sr_stash.address : sr_account.address

  const key: GenesisNodeKey = [
    address,
    address,
    {
      grandpa: ed_account.address,
      babe: sr_account.address,
      im_online: sr_account.address,
      parachain_validator: sr_account.address,
      authority_discovery: sr_account.address,
      para_validator: sr_account.address,
      para_assignment: sr_account.address,
      beefy: encodeAddress(ec_account.publicKey),
      aura: sr_account.address,
    },
  ];
```

File: src/chain-spec.ts [hunk 2]
```diff
@@ -114,86 +100,109 @@ export async function addBalances(specPath: string, nodes: Node[]) {
...
+    )} - ${decorators.magenta(sr_account.address)}`,
   );
 
   writeChainSpec(specPath, chainSpec);
 }
 
+/// Add node to staking
+export async function addParaCustom(specPath: string, node: Node) {
+  /// nop
+}
```
Comment (wirednkod) [inline src/chain-spec.ts:204]: Im not sure what this piece of code is. If is temp then maybe add a:
```suggestion
/// Add node to staking
export async function addParaCustom(specPath: string, node: Node) {
  /// nop
  console.info("Not implemented yet");
}
```

File: src/paras-decorators/moonbeam.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+  heath: "0x0d6dcaaef49272a5411896be8ad16c01c35d6f8c18873387b71fbc734759b0ab",
+  ida: "0x4c42532034540267bf568198ccec4cb822a025da542861fcb146a5fab6433ff8",
+  judith: "0x94c49300a58d576011096bcb006aa06f5a91b34b4383891e8029c21dc39fbb8b",
+};
+
+function specHaveSessionsKeys(chainSpec: ChainSpec) {
+  let keys = _specHaveSessionsKeys(chainSpec);
+
+  const runtimeConfig = getRuntimeConfig(chainSpec);
+  return keys || runtimeConfig?.authorMapping;
```
Comment (wirednkod) [inline src/paras-decorators/moonbeam.ts:41]:
```suggestion
  return keys || getRuntimeConfig(chainSpec)?.authorMapping;
```

File: src/paras-decorators/moonbeam.ts [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+function getAuthorityKeys(chainSpec: ChainSpec) {
+  const runtimeConfig = getRuntimeConfig(chainSpec);
+  if (runtimeConfig?.authorMapping) return runtimeConfig.authorMapping.mappings;
+}
```
Comment (wirednkod) [inline src/paras-decorators/moonbeam.ts:47]:
```suggestion
function getAuthorityKeys(chainSpec: ChainSpec) {
  return getRuntimeConfig(chainSpec)?.authorMapping?.mappings;
}
```

File: src/paras-decorators/moonbeam.ts [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+  const chainSpec = readAndParseChainSpec(specPath);
+
+  const { sr_account } = node.accounts;
+
+  let keys = getAuthorityKeys(chainSpec);
+  console.log(keys);
+  if (!keys) return;
+
+  keys.push(key);
+  console.log(keys);
```
Comment (wirednkod) [inline src/paras-decorators/moonbeam.ts:60]: remove non-needed console

File: src/paras-decorators/moonbeam.ts [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+async function addAuthority(specPath: string, node: Node, key: GenesisNodeKey) {
+  console.log(key);
+  const chainSpec = readAndParseChainSpec(specPath);
+
+  const { sr_account } = node.accounts;
+
+  let keys = getAuthorityKeys(chainSpec);
+  console.log(keys);
```
Comment (wirednkod) [inline src/paras-decorators/moonbeam.ts:56]: remove non-needed console

File: src/paras-decorators/moonbeam.ts [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+  keys.push(key);
+  console.log(keys);
+
+  console.log(
+    `\t👤 Added Genesis Authority ${decorators.green(
+      node.name,
+    )} - ${decorators.magenta(sr_account.address)}`,
+  );
+
+  console.log(chainSpec.genesis.runtime.authorMapping);
```
Comment (wirednkod) [inline src/paras-decorators/moonbeam.ts:68]: remove non-needed console

File: src/chain-spec.ts (hunk 2 above)
Comment (pepoviola) [reply src/chain-spec.ts]: This is just a `noop` allowing to customize by each `para`. Thanks!

**Comment by pepoviola** [comment]
@wirednkod I will merge this one 👍, thanks!!

**Review by wirednkod — APPROVED** [review]

**Pull request merged**